### PR TITLE
JWT Token feature

### PR DESCRIPTION
**Pull request by prasa7**
### Proposed changes in this pull request
This feature will give the capability to generate a JWT token instead oc a normal access token

-

### When should this PR be merged
ASAP

#### Code
This PR includes SQL changes as well as code changes.

#### Documentation
To use this feature the IdentityOAuthTokenGenerator should be changed to org.wso2.carbon.apimgt.keymgt.issuers.APIMTokenIssuer in identity.xml in the product distribution.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 1]
```diff
@@ -527,7 +528,8 @@ public static NativeObject jsFunction_mapExistingOauthClient(Context cx, Scripta
                 String applicationName = (String) apiData.get("applicationName", apiData);
 
                 String keyType = (String) apiData.get("keytype", apiData);
-                Map<String, Object> keyDetails = getAPIConsumer(thisObj).mapExistingOAuthClient(jsonString, userName, clientId, applicationName, keyType);
+                String tokenType = "DEFAULT";
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: In line 137 "DEFAULT_TOKEN_TYPE is already defined. We can remove this and use that instead.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 2]
```diff
@@ -3000,10 +3002,14 @@ else if (APIConstants.API_KEY_TYPE_SANDBOX.equals(tokenType)) { // if
             if (StringUtils.isBlank(callbackUrl)) {
                 callbackUrl = null;
             }
+            String applicationTokenType = application.getTokenType();
+            if(StringUtils.isEmpty(application.getTokenType())){
```
Comment (dinusha92) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Formatting issue here.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 3]
```diff
@@ -3000,10 +3002,14 @@ else if (APIConstants.API_KEY_TYPE_SANDBOX.equals(tokenType)) { // if
...
+                applicationTokenType = APIConstants.APPLICATION_TOKEN_TYPE;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java]: Is this correct? What are the possible values for token type? Not only JWT and Default?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/APISubscriptionInfoDTO.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
+/*
+*  Copyright (c) 2005-2011, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/APISubscriptionInfoDTO.java]: Incorrect year

Comment (bhathiya) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dto/APISubscriptionInfoDTO.java]: We can remove year.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/resources/dbscripts/h2.sql [hunk 5]
```diff
@@ -899,6 +899,7 @@ CREATE TABLE IF NOT EXISTS AM_APPLICATION (
     UPDATED_BY VARCHAR(100),
     UPDATED_TIME TIMESTAMP,
     UUID VARCHAR(256),
+    TOKEN_TYPE VARCHAR(100),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/resources/dbscripts/h2.sql]: Do we need 100? 10 not enough?

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+import org.wso2.carbon.apimgt.api.model.KeyManagerConfiguration;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.dto.JwtTokenInfoDTO;
+import org.wso2.carbon.apimgt.impl.dto.SubscribedApiDTO;
+import org.wso2.carbon.apimgt.impl.dto.SubscriptionPolicyDTO;
+import org.wso2.carbon.apimgt.impl.factory.KeyManagerHolder;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+
+import java.nio.charset.Charset;
+import java.util.*;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java]: Let's remove * imports.

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        String jwtHeader = null;
+
+        //if signature algo==NONE, header without cert
+        if (NONE.equals(signatureAlgorithm)) {
+            StringBuilder jwtHeaderBuilder = new StringBuilder();
+            jwtHeaderBuilder.append("{\"typ\":\"JWT\",");
+            jwtHeaderBuilder.append("\"alg\":\"");
+            jwtHeaderBuilder.append(getJWSCompliantAlgorithmCode(NONE));
+            jwtHeaderBuilder.append('\"');
+            jwtHeaderBuilder.append('}');
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java]: can't we use nimbus to build the header as well?

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                            .contains(userAttributeSeparator)) {
+                        StringTokenizer st = new StringTokenizer(claimVal, userAttributeSeparator);
+                        while (st.hasMoreElements()) {
+                            String attValue = st.nextElement().toString();
+                            if (StringUtils.isNotBlank(attValue)) {
+                                claimList.add(attValue);
+                            }
+                        }
+                        jwtClaimsSetBuilder.claim(claimURI, claimList.toArray(new String[claimList.size()]));
+                    } else if ("exp".equals(claimURI)) {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/token/APIMJWTGenerator.java]: Need a constant

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    public static JwtTokenInfoDTO getJwtTokenInfoDTO(Application application) {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        String userName = application.getOwner();
+
+        String tenantedUserName;
+        if (StringUtils.isEmpty(tenantDomain) || tenantDomain.equals("carbon.super")) {
+            tenantedUserName = userName;
+        } else {
+
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java]: extra new line

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            tenantedUserName = userName + "@" + tenantDomain;
+        }
+        APISubscriptionInfoDTO[] apis = new APISubscriptionInfoDTO[0];
+        try {
+            apis = ApiMgtDAO.getInstance().getSubscribedAPIsOfUserWithSubscriptionInfo(tenantedUserName);
+        } catch (APIManagementException e) {
+            e.printStackTrace();
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java]: print stack trace

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        JwtTokenInfoDTO jwtTokenInfoDTO = new JwtTokenInfoDTO();
+        jwtTokenInfoDTO.setSubscriber("sub");
+        jwtTokenInfoDTO.setEndUserName(userName);
+        jwtTokenInfoDTO.setContentAware(true);
+
+        List<SubscribedApiDTO> subscribedApiDTOList = new ArrayList<SubscribedApiDTO>();
+        for (APISubscriptionInfoDTO api : apis)
+
+        {
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/util/APIMTokenIssuerUtil.java]: not formatted

**Comment by Krishanx92** [comment]
Fix https://github.com/wso2/product-apim/issues/3326

**Pull request merged**